pkg/apis/core/validation/validation.go

```diff
@@ -6094,3 +6145,161 @@ func ValidateSpreadConstraintNotRepeat(fldPath *field.Path, constraint core.Topo
 	}
 	return nil
 }
+
+// validateServiceClusterIPsRelatedFields validates .spec.ClusterIPs,, .spec.IPFamilies, .spec.ipFamilyPolicy
+func validateServiceClusterIPsRelatedFields(service *core.Service) field.ErrorList {
+	// all validation is valid for non ExternalName services
+	if service.Spec.Type == core.ServiceTypeExternalName {
+		return field.ErrorList{}
+	}
+
+	allErrs := field.ErrorList{}
+	hasInvalidIPs := false
+	hasMisplacedNoneEmptyString := false
+
+	specPath := field.NewPath("spec")
+	clusterIPsField := specPath.Child("spec", "clusterIPs")
+	ipFamiliesField := specPath.Child("spec", "ipFamilies")
+	ipFamilyPolicyField := specPath.Child("spec", "ipFamilyPolicy")
+	// ipfamilies stand alone validation
+	// must be either IPv4 or IPv6
+	seen := sets.String{}
+	for i, ipFamily := range service.Spec.IPFamilies {
+		if !supportedServiceIPFamily.Has(string(ipFamily)) {
+			allErrs = append(allErrs, field.NotSupported(ipFamiliesField.Index(i), ipFamily, supportedServiceIPFamily.List()))
+		}
+		// no duplicate check also ensures that ipfamilies is dualstacked, in any order
+		if seen.Has(string(ipFamily)) {
+			allErrs = append(allErrs, field.Duplicate(ipFamiliesField.Index(i), ipFamily))
+		}
+		seen.Insert(string(ipFamily))
+	}
+
+	// must be maximum two
+	if len(service.Spec.IPFamilies) > 2 {
+		allErrs = append(allErrs, field.Invalid(ipFamiliesField, service.Spec.IPFamilies, "maximum allowed entries in ipfamilies is two"))
+	}
+
+	// multi family (dualstack) is only possible when the feature gate is enabled.
+	if !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) && len(service.Spec.IPFamilies) > 1 {
+		allErrs = append(allErrs, field.Invalid(ipFamiliesField, service.Spec.IPFamilies, "multiple ip families are only allowed when IPv6DualStack feature gate is enabled"))
+	}
+
+	// IPFamilyPolicy stand alone validation
+	if service.Spec.IPFamilyPolicy != nil /* nil is ok, defaulted in alloc */ {
+		// must have a supported value
+		if !supportedServiceIPFamilyPolicy.Has(string(*(service.Spec.IPFamilyPolicy))) {
+			allErrs = append(allErrs, field.NotSupported(ipFamilyPolicyField, service.Spec.IPFamilyPolicy, supportedServiceIPFamilyPolicy.List()))
+		}
+
+		// gate must be on
+		if !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) && *(service.Spec.IPFamilyPolicy) == core.RequireDualStack {
+			allErrs = append(allErrs, field.Invalid(ipFamilyPolicyField, service.Spec.IPFamilyPolicy, "RequireDualStack are only allowed when IPv6DualStack feature gate is enabled"))
+		}
+	}
+
+	// clusterIPs stand alone validation
+	// valid ips with None and empty string handling
+	// no duplication
+	seen = sets.String{}
+	for i, clusterIP := range service.Spec.ClusterIPs {
+		// duplicate check
+		if seen.Has(clusterIP) {
+			allErrs = append(allErrs, field.Duplicate(clusterIPsField.Index(i), clusterIP))
+		}
+		seen.Insert(string(clusterIP))
```

Comment:
this validation is here to ensure clearer error messages for user. duplicate vs `your IPs are not dualstack-ed`. It is the same practice we used for PodIPs.

Reply:
Sure, this error message is clearer in that case, but how in the world would that case ever happen? Is there some reason we think people might intentionally try to set a pair of duplicate ClusterIPs? Or that there is some plausible failure mode where they might do it accidentally?

Reply:
does how matter? I mean if it can happen, then we have to validate for it.

Reply:
We don't have to specifically validate "there are no duplicate IPs", because that case is already caught by the check that you don't have two IPs of the same family.

If we thought that trying to create a service with duplicate IPs was something that was going to happen a lot, then it might be useful to have a special error message for that case, to help users out. But if we don't have any reason to think it's going to be common, then we don't need to recognize it and call it out specially.

eg, the user could also try to set `clusterIP` to a hostname, and that would be invalid. But we don't have to specifically check `if validation.IsDNS1123Subdomain(clusterIP) { allErrs = append(allErrs, field.Invalid("cannot be a hostname") }` because (a) we don't actually expect people to do that, and (b) if they do it will get caught by the check that that clusterIP is a valid IP address anyway

Reply:
I will remove. But FYI we do pretty much the same in PodIPs